swift/apple/FirezoneKit/Sources/FirezoneKit/Views/UpdateNotification.swift

```diff
@@ -0,0 +1,245 @@
+//
+//  UpdateNotification.swift
+//  (c) 2024 Firezone, Inc.
+//  LICENSE: Apache-2.0
+//
+
+// Note: it should be easy to expand this module to iOS
+#if os(macOS)
+import Foundation
+import UserNotifications
+import Cocoa
+
+struct SemanticVersion: Decodable, Encodable, Comparable {
+  let major: Int
+  let minor: Int
+  let patch: Int
+
+  init(major: Int, minor: Int, patch: Int) {
+      self.major = major
+      self.minor = minor
+      self.patch = patch
+  }
+
+  // This doesn't conform to the full semver spec but it's enough for our use-case
+  private static func parse(versionString: String) -> (major: Int, minor: Int, patch: Int)? {
+      let components = versionString.split(separator: ".")
+      guard components.count == 3,
+            let major = Int(components[0]),
+            let minor = Int(components[1]),
+            let patch = Int(components[2]) else {
+          return nil
+      }
+      return (major, minor, patch)
+  }
+
+  init(from decoder: Decoder) throws {
+      let container = try decoder.singleValueContainer()
+      let versionString = try container.decode(String.self)
+
+      guard let parsed = SemanticVersion.parse(versionString: versionString) else {
+          throw DecodingError.dataCorruptedError(in: container,
+                                                 debugDescription: "Invalid SemVer string format")
+      }
+
+      self.major = parsed.major
+      self.minor = parsed.minor
+      self.patch = parsed.patch
+  }
+
+  static func from(string: String) -> SemanticVersion? {
+      guard let parsed = parse(versionString: string) else {
+          return nil
+      }
+      return SemanticVersion(major: parsed.major, minor: parsed.minor, patch: parsed.patch)
+  }
+
+  func encode(to encoder: Encoder) throws {
+      var container = encoder.singleValueContainer()
+      let versionString = "\(major).\(minor).\(patch)"
+      try container.encode(versionString)
+  }
+
+  static func < (lhs: SemanticVersion, rhs: SemanticVersion) -> Bool {
+      if lhs.major != rhs.major {
+          return lhs.major < rhs.major
+      }
+
+      if lhs.minor != rhs.minor {
+          return lhs.minor < rhs.minor
+      }
+
+      return lhs.patch < rhs.patch
+  }
+
+  static func == (lhs: SemanticVersion, rhs: SemanticVersion) -> Bool {
+      return lhs.major == rhs.major &&
+             lhs.minor == rhs.minor &&
+             lhs.patch == rhs.patch
+  }
+}
+
+private struct VersionInfo: Decodable {
+  let apple: SemanticVersion
+}
+
+private class UpdateNotifier: NSObject, UNUserNotificationCenterDelegate {
```

Comment:
```suggestion
private class NotificationAdapter: NSObject, UNUserNotificationCenterDelegate {
```

I don't know if there are any naming guidelines here in Swift. An "Adapter" could be a good naming choice because this class acts as a bridge between our code and the MacOS notification API.